### PR TITLE
Bs 182 SMS Error Code REST API

### PR DESCRIPTION
**Pull request by jaimecasero**
**What this PR does / why we need it**:
Exposes SMS error codes in REST API
**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes # Bs 182

**Special notes for your reviewer**:

### CODE REVIEW

File: restcomm/restcomm.application/src/main/webapp/WEB-INF/scripts/mariadb/init.sql [hunk 1]
```diff
@@ -249,7 +249,8 @@ direction VARCHAR(14) NOT NULL,
 price VARCHAR(8) NOT NULL,
 api_version VARCHAR(10) NOT NULL,
 uri MEDIUMTEXT NOT NULL,
-smpp_message_id MEDIUMTEXT
+smpp_message_id MEDIUMTEXT,
+error_code BIGINT
```
Comment (jaimecasero) [inline restcomm/restcomm.application/src/main/webapp/WEB-INF/scripts/mariadb/init.sql]: i would use a text based type instead

File: restcomm/restcomm.http/src/main/java/org/restcomm/connect/http/converter/AbstractConverter.java [hunk 2]
```diff
@@ -663,4 +664,35 @@ protected void writeFaxCapability(final Boolean faxCapable, final JsonObject obj
             object.addProperty("fax_capable", Boolean.FALSE);
         }
     }
+
+    protected void writeError(final Error error, final HierarchicalStreamWriter writer) {
```
Comment (hrosa) [inline restcomm/restcomm.http/src/main/java/org/restcomm/connect/http/converter/AbstractConverter.java]: If this is SMS-specific error, why did we create these methods in an generic converter?

File: restcomm/restcomm.sms/src/main/java/org/restcomm/connect/sms/smpp/SmppMessageHandler.java [hunk 3]
```diff
@@ -157,7 +157,10 @@ public void onReceive(Object message) throws Exception {
             if (sms == null) {
                 logger.warning("responseMessageId=" + smppMessageId + " was never received!");
             } else {
-                this.storage.getSmsMessagesDao().updateSmsMessage(sms.setSmppMessageId(null).setStatus(deliveryStatus));
+                sms.setSmppMessageId(null);
```
Comment (hrosa) [inline restcomm/restcomm.sms/src/main/java/org/restcomm/connect/sms/smpp/SmppMessageHandler.java]: CAREFUL!! each setter returns a brand new SMS object

File: restcomm/restcomm.sms/src/main/java/org/restcomm/connect/sms/smpp/dlr/provider/TelestaxDlrParser.java [hunk 4]
```diff
@@ -127,4 +165,21 @@ public Status parseRestcommStatus(String message) {
         return status;
     }
 
+    private Error parseRestcommErrorCode(String errCode) {
```
Comment (hrosa) [inline restcomm/restcomm.sms/src/main/java/org/restcomm/connect/sms/smpp/dlr/provider/TelestaxDlrParser.java]: Duplicated logic in ErrorCodeMapper, with difference that code is a String. Can't we maintain consistency in single place?

File: restcomm/restcomm.commons/src/main/java/org/restcomm/connect/commons/dao/Error.java [hunk 5]
```diff
@@ -0,0 +1,54 @@
+package org.restcomm.connect.commons.dao;
+
+/**
+ * @author mariafarooq
+ *
+ */
+public enum Error {
```
[the remark this answers is not in the document]
Comment (jaimecasero) [reply restcomm/restcomm.commons/src/main/java/org/restcomm/connect/commons/dao/Error.java]: agree,its even collisioning with existing class in JDK

File: restcomm/restcomm.sms/src/main/java/org/restcomm/connect/sms/smpp/dlr/provider/TelestaxDlrParser.java (hunk 4 above)
Comment (jaimecasero) [reply restcomm/restcomm.sms/src/main/java/org/restcomm/connect/sms/smpp/dlr/provider/TelestaxDlrParser.java]: not really duplicated.one maping is for DLR err codes,the other for submitResp.they are diff mapping as documented

File: restcomm/restcomm.http/src/main/java/org/restcomm/connect/http/converter/AbstractConverter.java (hunk 2 above)
Comment (jaimecasero) [reply restcomm/restcomm.http/src/main/java/org/restcomm/connect/http/converter/AbstractConverter.java]: agree this converter is getting loaded with specific entity details that will be rarely reused on diff converters,move to specific converter instead

File: restcomm/restcomm.sms/src/main/java/org/restcomm/connect/sms/smpp/SmppMessageHandler.java (hunk 3 above)
Comment (jaimecasero) [reply restcomm/restcomm.sms/src/main/java/org/restcomm/connect/sms/smpp/SmppMessageHandler.java]: ummm,right this is a bug for sure. I would say lets create a builder with copy constructor instead, this is very dangerous idiom we are using.

**Pull request merged**